### PR TITLE
Hmc

### PR DESCRIPTION
**Pull request by AdamI75**
I have completed my casperfpga HMC updates and I would like to merge into devel.

### CODE REVIEW

File: src/hmc.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
+import logging
+
+from memory import Memory
+
+LOGGER = logging.getLogger(__name__)
```
Comment (amishatishpatel) [inline src/hmc.py]: No need to instantiate this logging entity, just use the parent.logger

File: src/hmc.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    """
+    General HMC memory on the FPGA.
+    """
+    def __init__(self, parent, device_name, address, length_bytes, mezzanine_site, device_info=None, legacy_reg_map=True):
+        super(Hmc, self).__init__(name=device_name, width_bits=256, address=address, length_bytes=length_bytes)
+        self.parent = parent
+        self.device_info = device_info
+        self.device_name = device_name
+        self.mezz_site = mezzanine_site
+        self.reg_map = {# HMC Status Registers (LINK2)
```
Comment (amishatishpatel) [inline src/hmc.py:15]: Can the reg_map be moved to skarab_definitions.py?

File: src/hmc.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+                        'HMC_ERR_ABORT_NOT_CLEAR_HIGH_LINK3': 0xC8,
+                        'HMC_ERR_RSP_PACKET_LINK3': 0xCC,
+                        'HMC_ERRSTAT_LINK3': 0xD0,
+                        'HMC_CRC_ERR_CNT_LINK3': 0xD4,
+                        'HMC_STATUS': 0xD8}
+        # dictionary holding all HMC status information
+        self.hmc_status_list = {}
+        #dictionary holding all HMC revision information
+        self.hmc_revision_list = {}
+        LOGGER.debug('New Hmc %s' % self.__str__())
```
Comment (amishatishpatel) [inline src/hmc.py]: Print the device_name instead of the __str__ representation, so the user knows which HMC this is.

**Review by amishatishpatel — CHANGES_REQUESTED** [review]
Just a few changes, looks fine otherwise.

File: src/hmc.py (hunk 2 above)
Comment (tyronevb) [reply src/hmc.py:15]: Just a thought, maybe skarab_definitions.py isn't the best place for it? There could, for argument's sake, be 
SKARABs without HMCs. We'd rather, then, want to contain all HMC related info.

**Review by tyronevb — COMMENTED** [review]
Just having a quick look, things look fine.

File: src/hmc.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+                                                              self.reg_map['HMC_ERR_ABORT_NOT_CLEAR_LOW_LINK3']) + \
+                                                              (self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_ERR_ABORT_NOT_CLEAR_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_rsp_packet_link3'] = hex(self._wbone_rd(self.address +
+                                                           self.reg_map['HMC_ERR_RSP_PACKET_LINK3']))
+
+        self.hmc_status_list['hmc_errstat_link3'] = hex(self._wbone_rd(self.address + self.reg_map['HMC_ERRSTAT_LINK3']))
+
+        self.hmc_status_list['hmc_crc_err_count_link3'] = hex(self._wbone_rd(self.address +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:268]: I chose to insert this for readibility.

File: src/hmc.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+        self.hmc_status_list['hmc_err_abort_not_clr_link3'] = hex(self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_ERR_ABORT_NOT_CLEAR_LOW_LINK3']) + \
+                                                              (self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_ERR_ABORT_NOT_CLEAR_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_rsp_packet_link3'] = hex(self._wbone_rd(self.address +
+                                                           self.reg_map['HMC_ERR_RSP_PACKET_LINK3']))
+
+        self.hmc_status_list['hmc_errstat_link3'] = hex(self._wbone_rd(self.address + self.reg_map['HMC_ERRSTAT_LINK3']))
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:266]: Ignore

File: src/hmc.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+
+        self.hmc_status_list['hmc_run_lngth_bitflip_link3'] = hex(self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_RUN_LNGTH_BITFLIP_LOW_LINK3']) + \
+                                                              (self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_RUN_LNGTH_BITFLIP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_abort_not_clr_link3'] = hex(self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_ERR_ABORT_NOT_CLEAR_LOW_LINK3']) + \
+                                                              (self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_ERR_ABORT_NOT_CLEAR_HIGH_LINK3']) << 32))
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:261]: ignore

File: src/hmc.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+                                                      self.reg_map['HMC_ERR_ON_RX_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_run_lngth_bitflip_link3'] = hex(self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_RUN_LNGTH_BITFLIP_LOW_LINK3']) + \
+                                                              (self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_RUN_LNGTH_BITFLIP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_abort_not_clr_link3'] = hex(self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_ERR_ABORT_NOT_CLEAR_LOW_LINK3']) + \
+                                                              (self._wbone_rd(self.address +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:260]: ignore

File: src/hmc.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+                                                      self.reg_map['HMC_ERR_ON_RX_LOW_LINK3']) + \
+                                                      (self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_run_lngth_bitflip_link3'] = hex(self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_RUN_LNGTH_BITFLIP_LOW_LINK3']) + \
+                                                              (self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_RUN_LNGTH_BITFLIP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_abort_not_clr_link3'] = hex(self._wbone_rd(self.address +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:258]: ignore

File: src/hmc.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_on_rx_link3'] = hex(self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_LOW_LINK3']) + \
+                                                      (self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_run_lngth_bitflip_link3'] = hex(self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_RUN_LNGTH_BITFLIP_LOW_LINK3']) + \
+                                                              (self._wbone_rd(self.address +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:255]: ignore

File: src/hmc.py [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_LOW_LINK3']) + \
+                                                            (self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_on_rx_link3'] = hex(self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_LOW_LINK3']) + \
+                                                      (self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_run_lngth_bitflip_link3'] = hex(self._wbone_rd(self.address +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:253]: ignore

File: src/hmc.py [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+                                                      self.reg_map['HMC_RCVD_RESP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_tx_link_retries_link3'] = hex(self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_LOW_LINK3']) + \
+                                                            (self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_on_rx_link3'] = hex(self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_LOW_LINK3']) + \
+                                                      (self._wbone_rd(self.address +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:250]: ignore

File: src/hmc.py [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+                                                            self.reg_map['HMC_POISONED_PACKET_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_rcvd_resp_link3'] = hex(self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_RCVD_RESP_LOW_LINK3']) + \
+                                                      (self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_RCVD_RESP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_tx_link_retries_link3'] = hex(self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_LOW_LINK3']) + \
+                                                            (self._wbone_rd(self.address +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:245]: ignore

File: src/hmc.py [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+                                                            self.reg_map['HMC_POISONED_PACKET_LOW_LINK3']) + \
+                                                            (self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_POISONED_PACKET_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_rcvd_resp_link3'] = hex(self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_RCVD_RESP_LOW_LINK3']) + \
+                                                      (self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_RCVD_RESP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_tx_link_retries_link3'] = hex(self._wbone_rd(self.address +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:243]: ignore

File: src/hmc.py (hunk 2 above)
Comment (AdamI75) [reply src/hmc.py:15]: I am going to leave as is. I have discussed this with Wesley and he suggested that it is better practice that this gets moved to hmc.py. The whole idea of skarab_definitions.py was to accommodate the SKARAB BSP functions. The HMC is part of the DSP toolflow. Tyrone raises a valid point.

File: src/hmc.py [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_LOW_LINK3']) + \
+                                                            (self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_on_rx_link3'] = hex(self._wbone_rd(self.address +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:248]: ignore

File: src/hmc.py [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+                                                      (self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_run_lngth_bitflip_link3'] = hex(self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_RUN_LNGTH_BITFLIP_LOW_LINK3']) + \
+                                                              (self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_RUN_LNGTH_BITFLIP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_abort_not_clr_link3'] = hex(self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_ERR_ABORT_NOT_CLEAR_LOW_LINK3']) + \
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py]: Agreed. Fixed.

File: src/hmc.py [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+                                                            (self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_on_rx_link3'] = hex(self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_LOW_LINK3']) + \
+                                                      (self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_run_lngth_bitflip_link3'] = hex(self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_RUN_LNGTH_BITFLIP_LOW_LINK3']) + \
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py]: Agreed. Fixed.

File: src/hmc.py [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+                                                      (self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_RCVD_RESP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_tx_link_retries_link3'] = hex(self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_LOW_LINK3']) + \
+                                                            (self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_on_rx_link3'] = hex(self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_LOW_LINK3']) + \
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py]: Agreed. Fixed.

File: src/hmc.py [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+                                                            (self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_POISONED_PACKET_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_rcvd_resp_link3'] = hex(self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_RCVD_RESP_LOW_LINK3']) + \
+                                                      (self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_RCVD_RESP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_tx_link_retries_link3'] = hex(self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_LOW_LINK3']) + \
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py]: Agreed. Fixed.

File: src/hmc.py (hunk 1 above)
Comment (AdamI75) [reply src/hmc.py]: Agreed. Fixed.

File: src/hmc.py (hunk 3 above)
Comment (AdamI75) [reply src/hmc.py]: Agreed. Fixed.

File: src/hmc.py [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+        """
+        Read HMC revision.
+        :param
+        :return: self.hmc_revision_list - this is a dictionary containing HMC Vendor ID, HMC Product Revision,
+                                          HMC Protocol Revision and HMC Phy Revision
+        """
+        mezz_site = self.mezz_site + 1
+        # Reads back the Revisions and Vendor ID (Register Address: 0x2C0004) - Refer to page 4 of the Micron HMC
+        # Register Addendum data sheet for field mapping
+        hmc_revision = self.parent.transport.read_hmc_i2c(interface = mezz_site,
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:285]: ignore

File: src/hmc.py [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+                                                            (self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_on_rx_link3'] = hex(self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_LOW_LINK3']) +
+                                                      (self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_run_lngth_bitflip_link3'] = hex(self._wbone_rd(self.address +
+                                                              self.reg_map['HMC_RUN_LNGTH_BITFLIP_LOW_LINK3']) +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:254]: ignore

File: src/hmc.py [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+                                                      (self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_RCVD_RESP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_tx_link_retries_link3'] = hex(self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_LOW_LINK3']) +
+                                                            (self._wbone_rd(self.address +
+                                                            self.reg_map['HMC_TX_LINK_RETRIES_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_err_on_rx_link3'] = hex(self._wbone_rd(self.address +
+                                                      self.reg_map['HMC_ERR_ON_RX_LOW_LINK3']) +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:249]: ignore

File: src/hmc.py [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+                                                    self.reg_map['HMC_SENT_NP_LOW_LINK3']) +
+                                                    (self._wbone_rd(self.address +
+                                                    self.reg_map['HMC_SENT_NP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_sent_r_link3'] = hex(self._wbone_rd(self.address +
+                                                   self.reg_map['HMC_SENT_R_LOW_LINK3']) +
+                                                   (self._wbone_rd(self.address +
+                                                   self.reg_map['HMC_SENT_R_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_poisoned_packet_link3'] = hex(self._wbone_rd(self.address +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:233]: ignore

File: src/hmc.py [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+                                                   self.reg_map['HMC_SENT_P_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_sent_np_link3'] = hex(self._wbone_rd(self.address +
+                                                    self.reg_map['HMC_SENT_NP_LOW_LINK3']) +
+                                                    (self._wbone_rd(self.address +
+                                                    self.reg_map['HMC_SENT_NP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_sent_r_link3'] = hex(self._wbone_rd(self.address +
+                                                   self.reg_map['HMC_SENT_R_LOW_LINK3']) +
+                                                   (self._wbone_rd(self.address +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:230]: ignore

File: src/hmc.py [hunk 24]
```diff
@@ -0,0 +1,296 @@
...
+                                                   self.reg_map['HMC_SENT_P_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_sent_np_link3'] = hex(self._wbone_rd(self.address +
+                                                    self.reg_map['HMC_SENT_NP_LOW_LINK3']) +
+                                                    (self._wbone_rd(self.address +
+                                                    self.reg_map['HMC_SENT_NP_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_sent_r_link3'] = hex(self._wbone_rd(self.address +
+                                                   self.reg_map['HMC_SENT_R_LOW_LINK3']) +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:229]: ignore

File: src/hmc.py [hunk 25]
```diff
@@ -0,0 +1,296 @@
...
+                                                 self.reg_map['HMC_CTRL_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_sent_p_link3'] = hex(self._wbone_rd(self.address +
+                                                   self.reg_map['HMC_SENT_P_LOW_LINK3']) +
+                                                   (self._wbone_rd(self.address +
+                                                   self.reg_map['HMC_SENT_P_HIGH_LINK3']) << 32))
+
+        self.hmc_status_list['hmc_sent_np_link3'] = hex(self._wbone_rd(self.address +
+                                                    self.reg_map['HMC_SENT_NP_LOW_LINK3']) +
+                                                    (self._wbone_rd(self.address +
```
[the remark this answers is not in the document]
Comment (AdamI75) [reply src/hmc.py:225]: ignore

**Pull request merged**